### PR TITLE
 __intrinsic_assert

### PR DESCRIPTION
**Pull request by pmqtt**
Implementation of __intrinsic_assert.

Currently, needed for Optional(T!: Type)

I hope it's useful

### CODE REVIEW

File: explorer/testdata/assert/assert.carbon [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+// RUN:   %{FileCheck} --match-full-lines --allow-unused-prefixes=false %s
+// RUN: %{explorer} --parser_debug --trace_file=- %s 2>&1 | \
+// RUN:   %{FileCheck} --match-full-lines --allow-unused-prefixes %s
+// AUTOUPDATE: %{explorer} %s
+// CHECK: ASSERTION ERROR: {{.*}}/explorer/testdata/assert/assert.carbon:16: "HALLO WELT"
+
+package ExplorerTest api;
+
+fn Main() -> i32 {
+  __intrinsic_assert(true, "HALLO WELT");
```
Comment (geoffromer) [inline explorer/testdata/assert/assert.carbon]: I think we should try to avoid using intrinsics directly in user code, so let's wrap this in an `Assert` function in the prelude.

File: explorer/testdata/mixin/fail-circular-mixing.carbon [hunk 2]
```diff
@@ -13,7 +13,7 @@ __mixin M1 {
   fn F1[me: Self](x: Self) -> Self{
      return x;
   }
-// CHECK: COMPILATION ERROR: {{.*}}/explorer/testdata/mixin/fail-circular-mixing.carbon:[[@LINE+1]]: 'M3' has not been declared yet
+  // CHECK: COMPILATION ERROR: {{.*}}/explorer/testdata/mixin/fail-circular-mixing.carbon:[[@LINE+1]]: 'M3' has not been declared yet
```
Comment (geoffromer) [inline explorer/testdata/mixin/fail-circular-mixing.carbon]: This change and the ones in the other test files seem unrelated to this PR. Can we revert them?

File: explorer/common/error_builders.h [hunk 3]
```diff
@@ -33,6 +33,10 @@ inline auto RuntimeError(SourceLocation loc) -> ErrorBuilder {
   return ErrorBuilder("RUNTIME ERROR", loc.ToString());
 }
 
+inline auto AssertionError(SourceLocation loc) -> ErrorBuilder {
```
Comment (geoffromer) [inline explorer/common/error_builders.h]: I'm not sure if we need a new kind of error for this -- it seems like `RuntimeError` would work.

File: explorer/interpreter/type_checker.cpp [hunk 4]
```diff
@@ -2360,6 +2360,21 @@ auto TypeChecker::TypeCheckExp(Nonnull<Expression*> e,
...
           return Success();
+        case IntrinsicExpression::Intrinsic::Assert: {
+          if (args.size() != 2) {
+            return CompilationError(e->source_loc())
+                   << "__intrinsic_assert takes 2 arguments";
+          }
+          CARBON_RETURN_IF_ERROR(ExpectExactType(
+              e->source_loc(), "__intrinsic_assert argument 0",
+              arena_->New<BoolType>(), &args[0]->static_type(), impl_scope));
+          CARBON_RETURN_IF_ERROR(ExpectExactType(
```
Comment (geoffromer) [inline explorer/interpreter/type_checker.cpp]: We shouldn't need to require an exact type match here, since the interpreter applies implicit conversion to this argument, so `ExpectType` should work.

It would be nice to allow implicit conversions on the first argument as well. That would mean switching it to use `ExpectType` here, and adding a suitable `Convert` call in the interpreter.

File: explorer/testdata/assert/assert.carbon (hunk 1 above)
Comment (pmqtt) [reply explorer/testdata/assert/assert.carbon]: You are right!

File: explorer/testdata/mixin/fail-circular-mixing.carbon (hunk 2 above)
Comment (pmqtt) [reply explorer/testdata/mixin/fail-circular-mixing.carbon]: Hm, ofcourse! But I have only run update_checks.py...

File: explorer/common/error_builders.h (hunk 3 above)
Comment (pmqtt) [reply explorer/common/error_builders.h]: Yes RuntimeError would work. But I think RuntimeError is sent by the interpreter that something is wrong. With asserts it decides the user and that should be distinguished, or not ?

File: explorer/interpreter/type_checker.cpp (hunk 4 above)
Comment (pmqtt) [reply explorer/interpreter/type_checker.cpp]: Ok! Do you have an example which form of the call of __intrinsic_assert does not work now?

File: explorer/testdata/mixin/fail-circular-mixing.carbon (hunk 2 above)
Comment (geoffromer) [reply explorer/testdata/mixin/fail-circular-mixing.carbon]: Yeah, it looks like these files have gotten out of sync with the format that that script wants. I'll send a separate PR to fix them.

File: explorer/interpreter/type_checker.cpp (hunk 4 above)
Comment (geoffromer) [reply explorer/interpreter/type_checker.cpp]: I don't know of any types that are convertible to `bool` or `String` yet, so I can't easily concoct an example. I think we probably don't need to test this case yet, if that's what you're worried about; we just need to make the intent clear in the code.

File: explorer/common/error_builders.h (hunk 3 above)
Comment (geoffromer) [reply explorer/common/error_builders.h]: I believe `RuntimeError` is for run-time bugs in user code, and we use `CHECK` for bugs in the interpreter itself. An assertion failure indicates a bug in user code, so `RuntimeError` seems like a good fit.

Comment (pmqtt) [reply explorer/common/error_builders.h]: OK...!

File: explorer/interpreter/type_checker.cpp (hunk 4 above)
Comment (pmqtt) [reply explorer/interpreter/type_checker.cpp]: Done

File: explorer/common/error_builders.h (hunk 3 above)
Comment (pmqtt) [reply explorer/common/error_builders.h]: Done

File: explorer/testdata/mixin/fail-circular-mixing.carbon (hunk 2 above)
Comment (pmqtt) [reply explorer/testdata/mixin/fail-circular-mixing.carbon]: Done

File: explorer/testdata/assert/assert.carbon (hunk 1 above)
Comment (pmqtt) [reply explorer/testdata/assert/assert.carbon]: Done

**Review by geoffromer — APPROVED** [review]
